[STABLE ABI] Add mutable_data_ptr() and const_data_ptr() methods to torch::stable::Tensor.

pearu opened the pull request:
This ghstack is a prerequisite for porting torchaudio C++ extensions to use torch stable ABI, see https://github.com/pytorch/audio/issues/4074, https://github.com/pytorch/audio/issues/4075, https://github.com/pytorch/audio/issues/4076, https://github.com/pytorch/audio/issues/4077, https://github.com/pytorch/audio/issues/4078

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #161891
* #167772

pearu committed:
Update

[ghstack-poisoned]

pearu committed:
Update

[ghstack-poisoned]

pearu committed:
Update

[ghstack-poisoned]

pearu committed:
Update

[ghstack-poisoned]

mikaylagawarecki reviewed (commented):
This seems to make sense to me as a convenience utility and matches what's in TensorBase.h

Do add some lightweight tests (the file structure for testing can be seen in https://github.com/pytorch/pytorch/pull/160214/files)

pearu committed:
Update

[ghstack-poisoned]

pearu committed:
Update

[ghstack-poisoned]

pearu commented:
@mikaylagawarecki I have added tests for `mutable_data_ptr` and `const_data_ptr` Tensor methods.

@janeyx99 I have addressed the concern about semantics by introducing `aoti_torch_get_{mutable,const}_data_ptr` shim functions that call `{mutable,const}_data_ptr` methods of the `TensorBase` class so that stable Tensor `{mutable,const}_data_ptr` will have the corresponding semantics.

Please review.

pearu committed:
Update

[ghstack-poisoned]

pearu committed:
Update

[ghstack-poisoned]

pearu committed:
Update

[ghstack-poisoned]

janeyx99 commented on torch/csrc/stable/tensor_struct.h:
```diff
@@ -74,12 +74,30 @@ class Tensor {
   }
 
+  template <typename T>
+  T* mutable_data_ptr() const {
```
What's the difference between having the template version be on the implementation directly versus what we have in TensorBase.h:

```
  const void* const_data_ptr() const {
    return this->unsafeGetTensorImpl()->data();
  }

  void* mutable_data_ptr() const {
    return this->unsafeGetTensorImpl()->mutable_data();
  }
...
template <typename T, std::enable_if_t<!std::is_const_v<T>, int> = 0>
  const T* const_data_ptr() const;

  template <typename T, std::enable_if_t<std::is_const_v<T>, int> = 0>
  const std::remove_const_t<T>* const_data_ptr() const;

  template <typename T>
  T* mutable_data_ptr() const;
```
?

pearu replied:
This PR now exposes non-template methods `mutable_data_ptr` and `const_data_ptr`. The corresponding template methods (with proper type checking) will be implemented in a follow-up.

janeyx99 replied:
I think we should add the templating in the same PR to match the semantics of the nonstable version.

pearu replied:
> I think we should add the templating in the same PR to match the semantics of the nonstable version.

This is now done in this PR.

janeyx99 reviewed (commented):
I believe this is a convenience function, a "nice-to-have" over a "must-have", so we can take our time making sure that we land the right thing here. It is important to get this right as we want to minimize churn in a directory we label as stable.

pearu committed:
Update

[ghstack-poisoned]

pearu committed:
Update

[ghstack-poisoned]

pearu committed:
Update

[ghstack-poisoned]

pearu committed:
Update

[ghstack-poisoned]

pearu committed:
Update

[ghstack-poisoned]

pearu committed:
Update

[ghstack-poisoned]

pearu committed:
Update

[ghstack-poisoned]

pearu committed:
Update

[ghstack-poisoned]

pearu committed:
Update

[ghstack-poisoned]

mikaylagawarecki commented on torch/csrc/stable/ScalarType.h:
```diff
@@ -0,0 +1,50 @@
+#pragma once
+
+/*
```
Could you extend the comment to describe the rationale for this file (iiuc that the below mirrors  `AT_FORALL_SCALAR_TYPES_WITH_COMPLEX_AND_QINTS` and `toString` for ScalarType but the set of ScalarTypes supported by torch/csrc/stable is a subset of that supported by libtorch)

pearu replied:
The stable `toString` support is now in torch core, so this comment and `toString` implementation is obsolete here, I'll remove these.

mikaylagawarecki commented on test/cpp_extensions/libtorch_agnostic_extension/test/test_libtorch_agnostic.py:
```diff
@@ -14,11 +14,38 @@
 )
 
 
+def get_supported_dtypes():
```
do we have any hope of programmatically generating this?

mikaylagawarecki commented on test/cpp_extensions/libtorch_agnostic_extension/libtorch_agnostic/csrc/kernel.cpp:
```diff
@@ -476,14 +476,59 @@ void boxed_test_default_constructor(
+  stack[0] = from(res);
+}
+
+uint64_t get_template_any_data_ptr(Tensor t, Tensor r, bool mutable_) {
```
nit: could the ScalarType be passed directly? why do we need to use a tensor to carry it

pearu replied:
yes, passing ScalarType works now (after moving ScalarType to headeronly). I'll update tests!

mikaylagawarecki reviewed (commented):
I think this mostly makes sense to me and my comments are mostly nits/questions

To check my understanding this PR is mostly following the existing conventions in [TensorMethods.cpp](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/templates/TensorMethods.cpp?fbclid=IwY2xjawNc99VleHRuA2FlbQIxMQBicmlkETBhRjNHRG5BZEZIcjRTdHVzAR5JAPrn0XYLgLJnJn-Sn2do4mMsV5ezMJWDtEBXLmN9WEL09RhyvMMrGIeXUQ_aem_YaYcz2RarDfpZnsemQrKPw) to handle the `<T>` templates for `const/mutable_data_ptr<T>`, with the caveat that the version copied into `torch/csrc/stable` it's rewritten to only support types that `torch::stable::Tensor.scalar_type()` supports, is that right?

pearu committed:
Update

[ghstack-poisoned]

pearu committed:
Update

[ghstack-poisoned]

pearu committed:
Update

[ghstack-poisoned]

pearu committed:
Update

[ghstack-poisoned]

mikaylagawarecki reviewed (commented):
just one comment about qints, otherwise this looks good to go

pearu committed:
Update

[ghstack-poisoned]

pearu commented on torch/csrc/stable/tensor_inl.h:
```diff
@@ -33,4 +33,40 @@ inline Device Tensor::device() const {
+#define _PRED(S1, S2) S1 == S2
+#define _PRED_QINT(S1, S2) S1 == S2 || toUnderlying(S1) == S2
+#define DEFINE_CAST(T, name) DEFINE_DATA_PTR_CAST(T, name, _PRED)
+#define DEFINE_CAST_QINT(T, name) DEFINE_DATA_PTR_CAST(T, name, _PRED_QINT)
```
Done.

pearu committed:
Update

[ghstack-poisoned]

mikaylagawarecki reviewed (approved):
Thanks!

mikaylagawarecki commented on torch/csrc/inductor/aoti_torch/shim_common.cpp:
```diff
@@ -281,6 +281,24 @@ AOTITorchError aoti_torch_get_data_ptr(
   });
 }
 
+AOTITorchError aoti_torch_get_const_data_ptr(
```
@pearu actually, I am planning to merge a pretty big test refactor, in order to avoid too much hassle for you, let's merge this and I will fix this in a follow up

pearu replied:
> @pearu actually, I am planning to merge a pretty big test refactor, in order to avoid too much hassle for you, let's merge this and I will fix this in a follow up

Thanks!
